src/Paprika/Utils/RefCountingDisposable.cs

```diff
@@ -57,23 +73,65 @@ protected bool TryAcquireLease()
 
     private void ReleaseLeaseOnce()
     {
-        var value = Interlocked.Decrement(ref _leases);
+        // Volatile read for starting value
+        var current = Volatile.Read(ref _leases.Value);
+        if (current <= NoAccessors)
+        {
+            // Mismatched Acquire/Release
+            ThrowOverDisposed();
+        }
 
-        if (value == NoAccessors)
+        while (true)
         {
-            if (Interlocked.CompareExchange(ref _leases, Disposing, NoAccessors) == NoAccessors)
+            var prev = Interlocked.CompareExchange(ref _leases.Value, current - 1, current);
+            if (prev != current)
+            {
+                current = prev;
+                // Add PAUSE instruction to reduce shared core contention
+                Thread.SpinWait(1);
+                continue;
+            }
+            if (prev == 1)
```

Comment:
Magic const. This could be 
```suggestion
            if (prev == Initial)
```

Reply:
Changed